src/language-server/project/rover/DocumentSynchronization.ts

```diff
@@ -5,23 +5,30 @@ import {
   DidOpenTextDocumentNotification,
   DidCloseTextDocumentNotification,
   TextDocumentPositionParams,
+  Diagnostic,
+  NotificationHandler,
+  PublishDiagnosticsParams,
 } from "vscode-languageserver-protocol";
 import { TextDocument } from "vscode-languageserver-textdocument";
 import { DocumentUri, GraphQLProject } from "../base";
 import { generateKeyBetween } from "fractional-indexing";
 import { Source } from "graphql";
-import { findContainedSourceAndPosition } from "../../utilities/source";
+import {
+  findContainedSourceAndPosition,
+  rangeInContainingDocument,
+} from "../../utilities/source";
 
 export interface FilePart {
   fractionalIndex: string;
   source: Source;
+  diagnostics: Diagnostic[];
 }
 
 export function handleFilePartUpdates(
-  parsed: Source[],
-  previousParts: FilePart[],
-): FilePart[] {
-  const newParts = [];
+  parsed: ReadonlyArray<Source>,
```

Comment:
Changed all arrays to `ReadonlyArray` here because I want to prevent any accidental mutation. Should have been like this from the start.